### PR TITLE
feat: forward OSC 7 (working directory) to parent terminal

### LINKED ISSUE #898: Feature: Support for OSC 7 ANSI escape sequence
Related discussion: #892, #642

https://wezfurlong.org/wezterm/shell-integration.html : 

> OSC 7 Escape sequence to set the working directory

How could this be implemented?

### PR DESCRIPTION
**Pull request by marciogranzotto**
## Summary

When a shell runs inside Zellij, it emits OSC 7 (`\e]7;file://hostname/path\e\\`) to report its working directory. Current upstream uses this sequence for internal CWD tracking, but does not forward it to the parent terminal (Ghostty, iTerm2, Kitty, etc.).

This PR makes Zellij capture the raw OSC 7 URI per-pane and forward it to the parent terminal, re-emitting on pane/tab focus changes. This enables:

- **Cmd-click on relative file paths** (e.g. Ghostty's `resolvePathForOpening`, iTerm2)
- **Proxy icon** in macOS title bar
- **"New tab in same directory"** in the parent terminal
- **Tab title** showing current directory

Closes #898, relates to #3184, #3512, #3811, #3954

## Why naive forwarding doesn't work

Simply passing every pane's OSC 7 through to the parent terminal would be incorrect. Consider: pane A sends OSC 7 with `/home/a`, Zellij forwards it. Then pane B sends OSC 7 with `/home/b`, Zellij forwards it. When focus returns to pane A, **no new OSC 7 is emitted** — neither pane knows focus changed — so the parent terminal is now stale with `/home/b`.

The correct approach (implemented here): Zellij stores each pane's last-reported CWD, and on focus change re-emits the focused pane's CWD to the parent terminal.

## Design

- Store the **raw URI** on `Grid` exactly as the shell sent it — no parsing or normalization. This preserves hostname, percent-encoding, and scheme (`file://` vs `kitty-shell-cwd://`)
- Forward via the existing `post_vte_instructions` mechanism (already used for BEL forwarding) — no new IPC messages, proto changes, or client-side changes
- Track per-client what was last forwarded (`HashMap<ClientId, Option<String>>` on `Screen`), so non-mirrored multi-client sessions work correctly
- URIs with control characters are rejected to prevent escape sequence injection into the parent terminal
- `osc7_payload` is cleared on full terminal reset (RIS / `\ec`)

## Implementation

| File | Change |
|---|---|
| `zellij-server/src/panes/grid.rs` | Add `osc7_payload` field, extend the existing `b"7"` arm in `osc_dispatch` (joins `params[1..]` with `;`, rejects control chars), getter, clear on reset |
| `zellij-server/src/tab/mod.rs` | Add `osc7_payload()` to `Pane` trait with default `None` |
| `zellij-server/src/panes/terminal_pane.rs` | Implement `osc7_payload()` delegating to `self.grid` |
| `zellij-server/src/screen.rs` | Per-client `last_forwarded_osc7`, inject OSC 7 via `add_post_vte_instruction_to_client` during `render_to_clients`, cleanup on client disconnect |

## Edge cases

- **Pane never sent OSC 7:** Nothing forwarded — we don't fabricate a CWD
- **Multiple clients (non-mirrored):** Each client tracks its own last-forwarded state, so each parent terminal reflects its focused pane's CWD
- **Hostname/encoding/scheme:** All preserved from the original OSC 7 (raw string storage)
- **Terminal reset (`\ec`):** Clears stored `osc7_payload` to avoid stale data

## Scope

- Preserve upstream's internal CWD tracking, including the OSC 7 handling added in [#5063](https://github.com/zellij-org/zellij/pull/5063). Raw forwarding capture runs alongside it in the same handler.
- No configuration option needed — always-on, zero cost when shells don't emit OSC 7
- This is specifically OSC 7 forwarding, not a generic escape sequence passthrough

## Test plan

- [x] Unit test: OSC 7 URI stored from PTY bytes
- [x] Unit test: payload updates on subsequent OSC 7
- [x] Unit test: percent-encoded paths preserved
- [x] Unit test: BEL-terminated OSC 7 works
- [x] Unit test: control characters in URI rejected
- [x] Manual: cd /tmp in pane, cmd-click relative path in Ghostty - resolves correctly
- [x] Manual: switch between panes with different CWDs - parent terminal CWD updates
- [x] Manual: switch tabs - parent terminal CWD updates

## Rebase verification (2026-09-14)

- Rebased onto [upstream main](https://github.com/zellij-org/zellij/commit/232c9372) and removed AI co-author trailers from the feature commits.
- Integrated capture with upstream’s existing OSC 7 handler; regression assertions check that internal CWD updates and persistent raw-URI capture both work.
- `cargo test --locked --all-features -p zellij-server -- osc7 --nocapture`: 16 passed.
- `cargo xtask format --check`: passed.
- The manual checks listed above were performed before this rebase.

### CODE REVIEW

File: zellij-server/src/panes/grid.rs [hunk 1]
```diff
@@ -3464,6 +3472,23 @@ impl Perform for Grid {
...
+            // Join params[1..] with ";" to handle URIs containing semicolons
+            // (same pattern used by the b"0"|b"2" title handler above).
+            b"7" => {
+                if params.len() >= 2 {
+                    let uri = params[1..]
+                        .iter()
+                        .flat_map(|x| str::from_utf8(x))
+                        .collect::<Vec<&str>>()
+                        .join(";");
+                    if !uri.is_empty() && !uri.bytes().any(|b| b < 0x20 || b == 0x7f) {
```
[the remark this answers is not in the document]
Comment (marciogranzotto) [reply zellij-server/src/panes/grid.rs]: Good catch — C1 control characters (particularly U+009C / ST) could be interpreted by some parent terminals. Switched to `uri.chars().any(|c| c.is_control())` which covers both C0 and C1 ranges. Added a dedicated test for C1 rejection.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: zellij-server/src/panes/grid.rs [hunk 2]
```diff
@@ -3464,6 +3472,23 @@ impl Perform for Grid {
...
+            // Store the raw URI for forwarding to the parent terminal.
+            // Join params[1..] with ";" to handle URIs containing semicolons
+            // (same pattern used by the b"0"|b"2" title handler above).
+            b"7" => {
+                if params.len() >= 2 {
+                    let uri = params[1..]
+                        .iter()
+                        .flat_map(|x| str::from_utf8(x))
+                        .collect::<Vec<&str>>()
+                        .join(";");
```
[the remark this answers is not in the document]
Comment (marciogranzotto) [reply zellij-server/src/panes/grid.rs]: Fair point about silent segment dropping. This pattern matches the existing title handler (`b"0"|b"2"` at line 3286), but for URIs it's better to reject entirely on bad UTF-8. Switched to collecting into `Option<Vec<&str>>` via `.map(|x| str::from_utf8(x).ok()).collect()` — the entire OSC 7 is now discarded if any segment is invalid.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: zellij-server/src/screen.rs [hunk 3]
```diff
@@ -1983,7 +1985,33 @@ impl Screen {
...
+                    let last = self.last_forwarded_osc7.get(&client_id);
+                    if last != Some(&current_osc7) {
+                        if let Some(ref uri) = current_osc7 {
+                            output.add_post_vte_instruction_to_client(
+                                client_id,
+                                &format!("\x1b]7;{}\x1b\\", uri),
+                            );
+                            has_osc7_update = true;
+                        }
+                        self.last_forwarded_osc7.insert(client_id, current_osc7);
```
[the remark this answers is not in the document]
Comment (marciogranzotto) [reply zellij-server/src/screen.rs]: Good optimization. Now comparing using borrowed `Option<&str>` against `last.as_deref()` — the `to_owned()` allocation only happens when the value actually changes and needs to be stored.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by imsnif** [comment]
Hey @marciogranzotto - thanks for this and for your patience as well! The approach seems sound. I'd be happy to give it a closer look, but could you please remove the LLM commits and rebase from main? I'd rather not provide free ad space to commercial companies in our commit log unless they sponsor the project.

Thanks!

**Comment by marciogranzotto** [comment]
Thanks @imsnif! Removed the AI co-author trailers from the feature commits and rebased onto current main. During the rebase, I integrated the forwarding capture with the OSC 7 handler added for internal CWD tracking in [#5063](https://github.com/zellij-org/zellij/pull/5063), preserving both behaviors. All 16 OSC 7 tests and the workspace formatting check pass.

**Comment by imsnif** [comment]
Thanks for the super quick response @marciogranzotto ! I added tests for some edge cases as well as some integration tests to make sure this is tested end-to-end. Will merge once CI passes. Thanks again!
